### PR TITLE
Disturbance observer control for trajectory tracking

### PR DESCRIPTION
**Pull request by Jaeyoung-Lim**
This PR implements a disturbance observer for trajectory tracking.

This should improve on compensating for rotor drag / unmodeled disturbances

### CODE REVIEW

**Comment by Jaeyoung-Lim** [comment]
@kimnerd

File: geometric_controller/src/geometric_controller_node.cpp [hunk 1]
```diff
@@ -343,4 +367,31 @@ bool geometricCtrl::ctrltriggerCallback(std_srvs::SetBool::Request &req,
...
+}
+
+Eigen::Vector3d geometricCtrl::disturbanceobserver(Eigen::Vector3d pos_error, Eigen::Vector3d acc_setpoint){
+
+  Eigen::Vector3d acc_input, yq, yp, d_hat;
+  double dhat_max, dhat_min;
+  double control_dt = 0.01;
+
+  dhat_max = 10.0;
+  dhat_min = 0.0;
```
Comment (kimnerd) [inline geometric_controller/src/geometric_controller_node.cpp]: dhat의 saturation은 그냥 -10 to 10 이런식이면 됩니당. 그래서 그냥 편하게 min = -max로 두시면 될듯합니다.

File: geometric_controller/src/geometric_controller_node.cpp [hunk 2]
```diff
@@ -343,4 +367,31 @@ bool geometricCtrl::ctrltriggerCallback(std_srvs::SetBool::Request &req,
...
+
+  for(int i = 0; i < acc_input.size(); i++){
+    //Update dob states
+    p_.at(i)(0) = p_.at(i)(0) + p_.at(i)(1) * control_dt;
+    p_.at(i)(1) = (-a0(i) * control_dt / tau(i)) * p_.at(i)(0) + (1 - a1(i) * control_dt / tau(i)) + control_dt * acc_setpoint(i);
```
Comment (kimnerd) [inline geometric_controller/src/geometric_controller_node.cpp]: p_.at(i)(1) = (-a0(i) * control_dt / (tau(i))^2) * p_.at(i)(0) + (1 - a1(i) * control_dt / tau(i)) + control_dt * acc_setpoint(i); 인듯

File: geometric_controller/src/geometric_controller_node.cpp [hunk 3]
```diff
@@ -343,4 +367,31 @@ bool geometricCtrl::ctrltriggerCallback(std_srvs::SetBool::Request &req,
...
+    q_.at(i)(0) = q_.at(i)(0) + control_dt * q_.at(i)(1);
+    q_.at(i)(1) = q_.at(i)(1) + control_dt * ( (-a0(i)/std::pow(tau(i), 2))  * q_.at(i)(0) + (-a1(i)/tau(i)) * q_.at(i)(1) + pos_error(i));
+
+    //Calculate outputs
+    yp(i) = (a0(i) / pow(tau(i), 2)) * p_.at(i)(0);
+    yq(i) = (-a1(i)*a0(i) / std::pow(tau(i), 3))*q_.at(i)(0) + (std::pow(a0(i),2) / std::pow(tau(i), 4))*q_.at(i)(1) + a0(i) / pow(tau(i),2) * pos_error(i);
```
Comment (kimnerd) [inline geometric_controller/src/geometric_controller_node.cpp]: yq(i) = (-a1(i)*a0(i) / std::pow(tau(i), 3))*q_.at(i)(0) - (std::pow(a0(i),2) / std::pow(tau(i), 4))*q_.at(i)(1) + a0(i) / pow(tau(i),2) * pos_error(i); 인듯

**Comment by Jaeyoung-Lim** [comment]
https://youtu.be/cnvEQQcjZz8

Diverges when a circular trajectory is given

**Comment by Jaeyoung-Lim** [comment]
Successful hover! Seems like it works quite well!

There were some typos in the formulations in the code

[Circular trajectory tracking](https://youtu.be/bRMsM5HPsNw "Circular trajectory tracking")

**Comment by Jaeyoung-Lim** [comment]
As this branch is working. I am merging this.

We need more architectural improvements on running different controllers, but I believe this should be a seprate PR after the jerk feed forward controller is implemented

**Pull request merged**